### ISSUE TITLE
Quality handling not quite working as expected

### ISSUE THREAD

**Issue by lizaSB** [pymedusa/Medusa#2572]
**Branch/Commit:** 0.1.11
**OS:** Ubuntu 16.04
I have moved over from SickRage and have been attempting to get Medusa behaving similarly to how SickRage does, that is moving up the list of qualities until it finds the "Best Match" at which point it stops searching.

After a bit of searching, I found this wiki article which helped me arrive at my current settings: https://github.com/pymedusa/Medusa/wiki/Qualities-Changes

Notably:

> If you have qualities in BOTH lists, it will accept any quality in the Allowed list, and continue looking for a better quality in the Allowed list, until it finds one in the Preferred list.

Great, nice and clear. As I understand it, Medusa will keep bumping up quality in the Allowed list until it finds a single match in the preferred list, but only if we utilize both Allowed and Preferred qualities.

**What you did:** Added a show with the quality settings as such.
Allowed: 720p hdtv, 1080p web-dl
Preferred: 720p bluray

**What happened:** Medusa downloaded the 720p hdtv release as expected; however when the web-dl was released Medusa did not snatch. If I perform a search, medusa still does not snatch. I have to perform a manual search and select the desired release to get it to download 1080p web-dl

**What you expected:** Medusa to download the 720p hdtv release, followed by the 1080p release, eventually to be replaced by the preferred quality 720p bluray once released.

If I am mis-interpretting something please let me know. I searched as thoroughly as I could before posting this issue!

**Comment by labrys** [comment]
@lizaSB Actually it should only ever download at most 2 qualities, one from the allowed and then one from the preferred.  If it finds a preferred quality first, it stops a that.  There are plans to change this handling in the future but not any time soon.  Also this is the same handling as [SickRage](https://github.com/SickRage/SickRage) but not the same (afaik) as [SiCKRAGETV](https://github.com/SiCKRAGETV/SiCKRAGE).  We do have a feature that the others do not though.  You can manually select results for an episode using our Manual Snatch feature and choose whichever release(s) you would like to download.

**Comment by lizaSB** [comment]
@labrys, so the quote I selected from the wiki is incorrect then?

Having come over from SickRage (not SiCKRAGETV), the quality moves up the Preferred list until it settles on a final. This is how I've been using it for some time and it works as expected. It has been since migrating to Medusa that I've been attempting to recreate this behavior without success.

For many shows I simply want 720p hdtv to be trumped by 720p web - that case is easy and is working as expected. However, for the shows that I would prefer BluRay for (very few), web-dl is the best interrim quality.

The manual search is nice, I admit this is part of my reason for giving this fork a shot; however, manually searching for the preferred quality each week becomes tiresome when other forks are already working as expected.

**Comment by fernandog** [comment]
@lizaSB are you using latest version?

Edit show qualities. A description of it will appear below the quality lists

**Comment by lizaSB** [comment]
@fernandog yes:
`cur_commit_hash = a0116154f262786d6f5274459cb8f9cdb23012cc`
Which aligns with the head of master

I have been reading under show qualities and it helped me understand why the "Preferred" quailty was no longer working as expected. I think the major issue is that the documentation of the quality settings are largely misleading. It still states that it will work up the Preferred list which is how it operates in SickRage. Then the Qualities-Changes sub page states the opposite, that it will work up the "Allowed" until it finds a "Preferred" match, when in reality Medusa finds one allowed and one preferred then stops listening entirely.

>Preferred Quality allows you to select a second set of Qualities that instructs Medusa to download any successively better matched Quality

Perhaps I am a minority here, but documentation is where I always search first to understand how things are supposed to work :wink: If this is indeed the intended behavior and not a bug and not to be changed any time soon, I can update the documentation.

**Comment by labrys** [comment]
@fernandog I noticed that description isn't as clear as it could be either.

@lizaSB I agree the documentation needs to be correct.

Ok so just to verify in most cases you want:

`720p Web > 720p HDTV` 

 but for some shows you want

`720p Blu-Ray > 720p Web > 720p HDTV`

Correct?

The closest configuration for the second is to do one of the following:

OPTION A:
![image](https://cloud.githubusercontent.com/assets/1867464/24778877/eec83542-1afa-11e7-8c06-97760a69b1ea.png)

OPTION B:
![image](https://cloud.githubusercontent.com/assets/1867464/24778902/1784a8bc-1afb-11e7-94e1-8447e062f641.png)

As mentioned this at most will only download 2 qualities.  The first download could be any of the three qualities the second would be an upgrade.

For option A it would upgrade an `HDTV` to `Web` or `BluRay`.
For option B it would upgrade an `HDTV` or `Web` to `BluRay`.

You can tweak this to improve the behaviour.  You may already be familiar with `Required` and `Ignored` words from SickRage. 

With required words if the word **is not** in the release name the release will not be selected.
With ignored words if the word **is** in the release name the release will not be selected.

The problem here is that those features are too rigid for your needs.  For example add something like `Web` to `Required` and you would never get `HDTV` or `BluRay`. 

To overcome this limitation, we have added another Medusa exclusive feature to global search settings `Preferred` and `Undesired` words.  With `Preferred` words a release with those words would be preferred over one without, but if no releases had a preferred word, it would still make a selection.  By the same token, if a release has an undesired word, it would select a different release first, but would not prevent it from being selected if there were no other releases.

#### TLD; DR
Try option B above and add `Web` to `Preferred words` (under Gear -> Search Settings -> Episode Search tab, Search Filters at the bottom)

![image](https://cloud.githubusercontent.com/assets/1867464/24779702/f6a4ee40-1aff-11e7-98a5-856be328a82e.png)

**Comment by labrys** [comment]
@lizaSB if you would like to discuss the reasons for the current search pattern and how it evolved, feel free to join us in IRC on Freenode #pymedusa available from within Medusa under `Tools` -> `IRC`

**Comment by lizaSB** [comment]
@labrys thank you for the extensive response.

The issue is that hdtv will always download first because 95% of the time, HDTV is released before WEB. Sometimes I watch the episode quickly and HDTV is all that's available, but often I wait for the better version to watch.

Knowing this, my flow will change to this: I will just set 720p hdtv to allowed and WEB/BluRay to preferred. The BluRay releases I want to download are so few and far between that manual searching for those would be much less of a hassle than manually snatching all the individual web episodes for the same releases. It's not perfectly ideal, but I do find myself preferring the little extra features in this fork :slightly_smiling_face: 

I appreciate your responses to this issue, and since we can all agree that documentation should be correct, I will update it there and consider this issue closed.

**Comment by labrys** [comment]
@lizaSB In that case you could select Option A and add `bluray` to preferred words.  Its not quite as bulletproof since bluray can be `Blu-Ray`, `BluRay`, `BD` etc, but it could reduce the number of manual searches required.

**Comment by gitowok** [comment]
@niayh  and anyone else wanting this behaviour back, I've taken the time to work out how to do it, and thought I'd document it here, as this is the most relevant google result.

This hack patches medusa to always upgrade to a higher quality in the allowed list (even if the new quality is not in the preferred list)

ie given:
Allowed: 720p WEB-DL, 1080p WEB-DL, 1080p BluRay
Preferred: 1080p BluRay

If old quality is 720p WEB-DL and a 1080p WEB-DL is found, medusa *will* download the upgrade.
If it finds a 1080p BluRay, it upgrades the quality and stops searching forever, as is normal behaviour.


medusa/common.py,  roughly line 708, find:
```
            # Replace if preferred quality
            if new_quality in preferred_qualities:
                return True, 'New quality is preferred. Accepting new quality'

            return False, 'New quality is same/lower quality (and not preferred). Ignoring new quality'
```
change to:
```
            # Replace if better (even if new quality is only in allowed list)
            if new_quality > old_quality:
                return True, 'New quality is better, and in allowed or preferred. Accepting new quality'

            return False, 'New quality is same/lower quality. Ignoring new quality'
```

to any devs listening - it would be wonderful to have this as a switchable preference!

Hope this helps someone

**Comment by gitowok** [comment]
@labrys  -  This patch addresses quality upgrades in the allowed list that are found before the preferred quality is available. Medusa by default ignores any upgrades in quality unless its the preferred quality. I'll give a clearer example: 

Given a **1080p BluRay** preferred quality, but **everything** in allowed quality:

- an SDTV is found - medusa downloads it.
- a 720p HDTV is found - medusa by default will ignore it (**but will download it with this patch**)
- a 1080p HDTV is found - medusa by default will ignore it (**but will download it with this patch**)
- a 1080p WEB-DL is found - medusa by default will ignore it (**but will download it with this patch**)
- a 1080p BluRay is found - medusa will download it and stop searching.

This is the behaviour the OP wanted, and myself also.

**Comment by gitowok** [comment]
For reference, and to bring complete closure here, this is now entirely fixed upstream as of latest release 0.2.6. Please see #4511 and #4514
